src/Hazel/Var.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+-- | An adaptation of the locally nameless representation (See Charguéraud's
+-- "The Locally Nameless Representation" for more information) which uses 2D
+-- de Bruijn variables.
+
+module Hazel.Var where
+
+import Control.Lens (Iso', Traversal', iso)
+import Data.Word (Word32)
+
+import qualified Data.Text as T
+
+-- | The distance in binding levels between a variable and its binder. The
+-- "first dimension" in our 2D de Bruijn scheme.
+newtype Depth
+  = Depth { _depth :: Word32 }
```

Comment:
One devils advocate idea is to use Natural as exported by Numeric.Natural (I know I've also articulated for Word32 :) )

If you look at the way integer and natural are defined they both provide fast paths when the math fits in a native word / int, and on a dafter note Natural is the only unsigned integer ish type in base that errors when you try to go below zero.  Fixing / clarifying that is part of my 2020 hopes ;)

On the flip side, the extra casing and indirection may or may not impact performance since variables are like the only things languages manipulate and I think we all agree that no programs will have billions or even millions of variables.  

Pay this no mind. But figured I should mention it while it was in my head